### ISSUE TITLE
Build tooling and package interface improvements for v3

### ISSUE THREAD

**Issue by friendlyanon** [SFML/SFML#2079]
Since v3 is around the corner and it's a major version increase, I think it would be time to revamp the build tooling a little bit and follow proper library interface etiquette.

I'm opening this issue is just to spitball some ideas and enumerate wishes regarding what could be better.

One immediate thing I noticed is the presence of binary blobs in source control. This is some icky stuff that can be replaced with package managers easily. Conan and vcpkg are pretty much the only options here. Conan would be nice, because it has a `deploy` generator where the dependency graph can be installed to a staging prefix, but it requires more upfront work with setup, because Conan v2 no longer tries to guess details of the system. vcpkg requires less work upfront, since it just plugs into the CMake configure process using its toolchain script, but I'm not sure what's the proper way is to deploy with it (just copy the `vcpkg_installed` folder from build dir?).   
Either way, neither requires hardcoding details into the build scripts nor coupling, which means that users, developers and package managers (the people) can freely plug whatever they want to into the build. Both allow transparent integration with little to no work required.

There is tons of duplicate code of things that CMake already provides as a built-in, such as the setting of the CRT type on Windows. In fact, what's in the CML file right now if insufficient for providing the debug runtimes. CMake already has an alternative for this, so this code requires no new code, only some documentation in the install/build docs pointing people towards it.

SFML's install interface could also use improvements. The CMake package currently provided is essentially broken. Some details about the broken-ness can be found [here](https://github.com/friendlyanon/cmake-init-sfml-example/blob/master/CMakeLists.txt#L47-L62). As an aside, this is also what prompted #2052, which is just a partial fix, but the commit in that PR also contains the ideal UX, as in that would be all the code ever needed to use SFML in CMake.

The `cmake/Macros.cmake` file could (should) also be sacked, as it does a lot of things to break (sub)system level conventions and it again duplicates a lot of things that CMake provides using built-in variables. I'm very much not a fan of projects overriding user requirements and providing custom magic suffixes like `-static` for static libraries, when the users could make that consciuous decision themselves with no costs to bear for every downstream user. Most of these things should just be documented in the install/build document for users and recommending them to use CMake, since that takes care of all these little differences. Looking at the [Getting Started](https://www.sfml-dev.org/tutorials/2.5/#getting-started) section for example shows no usage with CMake right now, which stinks imo.  
Just as an aside, but I think projects should have and maintain a state-of-the-art (wrt the project itself) repository with CI that shows some basic usage for a quick cheat sheet. Doing this would reveal some deficiencies in the interface the project provides. For example, I tried setting up a [project](https://github.com/friendlyanon/cmake-init-sfml-example) to see how SFML rolls and from my earlier mention it's clear that things aren't as good as they could be.

I'm also not sure if you are aware, but right now the build scripts are setup in a way that a minor version increase on *nix systems is a breaking change (https://github.com/SFML/SFML/blame/9010e7ba381d3b07a3af85f26a974b55dc07cfba/cmake/Macros.cmake#L142), while the library advertises itself as following semantic versioning (https://github.com/SFML/SFML/blame/9010e7ba381d3b07a3af85f26a974b55dc07cfba/cmake/Macros.cmake#L455).

**Comment by ChrisThrasher** [comment]
I'm glad to see you're back thinking about how we can improve the SFML build! 

---

> One immediate thing I noticed is the presence of binary blobs in source control. 

I'd say this is the top build-related problem. https://github.com/SFML/SFML/issues/1991 already touches on some of these things. It's something I'm really motivated to improve. I've spoken to the SFML maintainers about it and they're generally onboard with removing the precompiled binaries, at least for all non-Windows OSes.

> There is tons of duplicate code of things that CMake already provides as a built-in,

I've started chipping away at this. As you might imagine this is largely because the build hasn't been significantly changed since much of it was written many years ago so it's still using CMake practices circa 2018. You'll likely find no resistance if you submit PRs replacing existing CMake code with built-in functionality. Feel free to raise the minimum CMake version requirement if needed.

> The cmake/Macros.cmake file could (should) also be sacked

I somewhat agree. I'm not a fan of adding custom abstractions to the build, but currently I don't know of a better alternative to these macros. If these macros were significantly simplified then it would be reasonable to manually inline them.

> I'm also not sure if you are aware, but right now the build scripts are setup in a way that a minor version increase on *nix systems is a breaking change 

That should be an easy fix. I'll submit a PR for that unless you would rather take credit for it. Probably just an oversight. This was introduced way back in 2014 in 1c46ec7c37702a01e8397d63c3e3a0b9e70ff87b.

**Comment by friendlyanon** [comment]
> Feel free to raise the minimum CMake version requirement if needed.

I don't plan to and from the cursory look I took, it'd be wholly unnecessary. 3.8 is just fine, since that's when proper standard version selection was added using compile features ([[1]](https://cmake.org/cmake/help/latest/manual/cmake-compile-features.7.html#requiring-language-standards) [[2]](https://cmake.org/cmake/help/latest/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#high-level-meta-features-indicating-c-standard-support)), which SFML makes proper use of.

**Comment by friendlyanon** [comment]
I'm not sure if there were/are discussions taking place somewhere else, but I feel like the silence here just means there are no specifics I should take care of and I should just chew through this, then the details will be hashed out in the PR.

**Comment by ChrisThrasher** [comment]
There aren't any ongoing discussions and the discussion in #1991 hasn't been active in a while. The most helpful thing would be to start submitting PRs so we can see exactly what improvements you have in mind.

**Comment by friendlyanon** [comment]
What I have in mind is something similar to what I have in cmake-init, which is clear separation of code that is only consumer facing and developer facing. This isn't a new idea, a "developer mode" is something I have seen in a few stray projects as well. This developer mode would then be complemented with a separate developer facing build and contribution documentation and CMake presets.

The biggest issue I'll have to figure out is which package manager to involve with this project, see my points in OP.

I'll probably have something to look at during the weekend.

**Comment by Bambo-Borris** [comment]
> What I have in mind is something similar to what I have in cmake-init, which is clear separation of code that is only consumer facing and developer facing. This isn't a new idea, a "developer mode" is something I have seen in a few stray projects as well. This developer mode would then be complemented with a separate developer facing build and contribution documentation and CMake presets.
> 
> The biggest issue I'll have to figure out is which package manager to involve with this project, see my points in OP.
> 
> I'll probably have something to look at during the weekend.

Is there any reason this concept of developer mode vs consumer mode CMake is only seen on stray projects? Also do you have links to any examples where you've seen this used before that could be helpful to understand fully what you mean?

**Comment by ChrisThrasher** [comment]
I'm particularly excited to see the ways you think we can replace existing CMake code with built-in CMake functionality. Those kinds of PRs would get quickly accepted and likely full support from me.

**Comment by friendlyanon** [comment]
> Is there any reason this concept of developer mode vs consumer mode CMake is only seen on a few stray projects? 

I guess this sort of thing falls in the category of unknown unknowns. People are simply not aware that something like this could be done.

A developer mode allows the build process to be simpler for consumers, since they only get the code that actually contributes to defining the build and usage requirements of a project, while the developers can commit all the "attrocities" they want on the code path exclusive to them. This separation also makes it easier to make a project friendlier for users who vendor projects, since you have less to worry about.

Craig Scott also hinted at a few things that are necessary to be friendlier in that regard in [CppCon his talk](https://www.youtube.com/watch?v=m0DwB4OvDXk), but he only touched upon the install components, which is just one of the many things that need to be done.

> Also do you have links to any examples where you've seen this used before...

No, sorry. I have not taken note of those, but I know that even Jason Turner's starter template has started making use of such a concept https://github.com/aminya/project_options/commit/789f486930ee00468f3054a579042c821a5c5e85

> ...that could be helpful to understand fully what you mean?

https://github.com/friendlyanon/cmake-init/wiki/Examples naturally, all of these do that. The first three are always kept in sync with upstream, so those should reflect the "state-of-the-art" better regarding documentation differences. You can experiment with them and ask further questions if you need more details.

**Comment by ChrisThrasher** [comment]
Are you still interested in submitting PRs for any of the smaller scope fixes you have in mind? I'm still really interested in ways we can replace existing CMake code with built-in functionality from CMake <= 3.16.

**Comment by friendlyanon** [comment]
Yes. I started researching the various platforms SFML supports and slowly discovering what is actually needed and what isn't, but this week has been and will be quite busy for me, so I haven't really had the time to do much programming so far.

I have since realized that because SFML also builds Objective-C++ code, it must require CMake 3.16 as a minimum. Hacking around the lack of support for that language for prior versions feels like a fool's errand to me with how [trivial it is to keep CMake up to date](https://alexreinking.com/blog/how-to-use-cmake-without-the-agonizing-pain-part-1.html#:~:text=Picking%20a%20CMake%20Version).

**Comment by ChrisThrasher** [comment]
> I have since realized that because SFML also builds Objective-C++ code, it must require CMake 3.16 as a minimum

Not a problem. [CI already passes while requiring CMake 3.16](https://github.com/ChrisThrasher/SFML/actions/runs/2248240678) so it should be no problem to raise the requirement to that version. I don't think anybody wants to add hacks just to keep that 3.8 requirement in place.

**Comment by eXpl0it3r** [comment]
One of the upmost important points in all of this is, that SFML remains buildable without requiring additional third-party tooling. SFML is used a lot in education, on school PCs, in limited environments etc. if we start requiring people to additionally install python + conan + vckpkg + ... just to build SFML, we're not doing anyone a favor.

I'm on board if we simply add support for package managers, so that they neatly integrate and for those users who make use of them, life can be easier.

As such I'd envision a build system that can work with system-wide installed dependencies (e.g. through Linux package manager), C++ package manage (vcpkg / conan) or if none of them exist, it would FetchContent (or similar) get the dependency source and build them on-the-fly.

> > I have since realized that because SFML also builds Objective-C++ code, it must require CMake 3.16 as a minimum
> 
> Not a problem. [CI already passes while requiring CMake 3.16](https://github.com/ChrisThrasher/SFML/actions/runs/2248240678) so it should be no problem to raise the requirement to that version. I don't think anybody wants to add hacks just to keep that 3.8 requirement in place.

Whether CI passes is only insofar an indicator, as to whether our pipeline would break, the more important point is to see what's available on which OS through the "normal" installation paths, i.e. what versions do the most popular Linux distros ship (see [Repology](https://repology.org/project/cmake/versions)). And from there on it's then really just the max version that's required, but also the min version that necessary - no need to update to a version of which we don't use any of the features.

But it seems that CMake 3.16 is okay to use. ✔️ 

> The `cmake/Macros.cmake` file could (should) also be sacked

It's kind of funny how the link to Jason's repo regarding developer mode, then defines yet another custom option macro... 😄

But I also think, we don't really need to maintain our own versions of CMake functions. Some of them were needed due to the lack of CMake support. On the other hand, if we have to copy-paste x-times the same length bit of code. It might make sense to extract this into a macro, same as you would move duplicated code into a function.

> A developer mode allows the build process to be simpler for consumers, since they only get the code that actually contributes to defining the build and usage requirements of a project, while the developers can commit all the "attrocities" they want on the code path exclusive to them. This separation also makes it easier to make a project friendlier for users who vendor projects, since you have less to worry about.

I'm a bit on the fence about this. For one it feels like SFML's CMake system has a lot more urgent needs, than this. As such I'd recommend to first tackle the more important issues and when we're happy with the new  setup, we can consider, which parts of the "interface" should be hidden to developer mode.
Additionally, I can also see how this can add many more if-branches, depending on how it's setup, which might make the CMake code a lot harder to read again.

> I'm also not sure if you are aware, but right now the build scripts are setup in a way that a minor version increase on \*nix systems is a breaking change ([https://github.com/SFML/SFML/blame/9010e7ba381d3b07a3af85f26a974b55dc07cfba/cmake/Macros.cmake#L142](https://github.com/SFML/SFML/blame/9010e7ba381d3b07a3af85f26a974b55dc07cfba/cmake/Macros.cmake#L142)), while the library advertises itself as following semantic versioning ([https://github.com/SFML/SFML/blame/9010e7ba381d3b07a3af85f26a974b55dc07cfba/cmake/Macros.cmake#L455](https://github.com/SFML/SFML/blame/9010e7ba381d3b07a3af85f26a974b55dc07cfba/cmake/Macros.cmake#L455)).

The understanding of a breaking change is different between [Semantic Versioning](https://semver.org/) and what Linux systems/packaging. Sem version relates to API-level changes, while Linux packages usually also consider ABI-level changes. As such it's *not* an API breaking change if a function gets an additional, optional parameter, but *it is* an ABI breaking change, thus requires a bump in the SOVERSION. For more details see [the discussion](https://github.com/SFML/SFML/pull/745#issuecomment-65322744) with the Debian package maintainer @jcowgill.

---

To summarize:

- I'm happy to accept pretty much any CMake improvements.
- We can't accept that SFML becomes only buildable with package managers.
- I suggest to start with low-hanging fruits (e.g. proper use of CMake-options with ON/OFF or similar) and ensure that we don't get huge diffs

**Comment by friendlyanon** [comment]
Lock-in with Conan and vcpkg takes extra effort, meanwhile the simplest way of just using `find_package()` is universal, transparent and It Just Works (tm)... unless a dependency does not support clients using CMake, but then your best option is to simply fall back to manually `find_*()`ing things anyway.  
vcpkg's manifest mode in particular integrates seamlessly with the idiomatic CMake mechanisms using their toolchain file. Conan has generators that generate CMake package configs that `find_package()` works with.  
If you look at what I have been doing recently you can see that I'm no fan of lock-in or any deviations from the idiomatic CMake mechanisms.

> FetchContent (or similar) get the dependency source and build them on-the-fly.

I'm very much not a fan of vendoring in any form, but this is also doable using just `find_package()` by anyone who wishes to provide dependencies in such a way. Due to the flexibility of `find_package()`, one can make a find module visible to the command's module mode search procedure, where the script does use FetchContent. This is also I have a demonstration for: https://github.com/friendlyanon/cmake-init-fetchcontent  
Again, this is completely transparent here, the project code is not aware of how the dependency is made available, all it knows is that is must `find_package()` and the actual work is deferred to whatever that command finds.

> But it seems that CMake 3.16 is okay to use.

With CentOS being dead, I think the new measure for "whatever version my system distribution ships" can be Ubuntu LTS, which ships with 3.16, so that's no big issue. Then again, the CMake binaries are extremely portable and just extracting the archives gives you a fully functional CMake installation.

> It's kind of funny how the link to Jason's repo regarding developer mode, then defines yet another custom option macro

It was just more about another instance which I was aware of that makes the distinction between "the default user path" and "developer mode path". I have no comments on the quality of the project otherwise.

> It might make sense to extract this into a macro, same as you would move duplicated code into a function.

Yes, things like setting up unmarked library targets with the correct properties is one kind of repetition that I have already made a function for on my WIP branch. There is also a function for setting the sources for a target and setting their source group, just so the code defining a target's requirements can drop using variables. Some abstractions are useful.

> For one it feels like SFML's CMake system has a lot more urgent needs, than this. As such I'd recommend to first tackle the more important issues and when we're happy with the new setup, we can consider, which parts of the "interface" should be hidden to developer mode.

Being able to make the distinctions easier is a natural byproduct of a thorough shaving of the code. I have already chucked quite a bit out the window, since a lot of it is simply redundant with higher CMake versions, or simply there was no need for it to begin with.  
The usual dev-mode only code is tests, but any code that does not contribute to defining the requirements of the project are also candidates. Something I can think of here is setting the property that changes the startup project in Visual Studio. It's a nice-to-have, but only for someone who actually develops this project.

> The understanding of a breaking change is different between [Semantic Versioning](https://semver.org/) and what Linux systems/packaging.

Well, I'm still on the opinion that having the full major.minor version represent the ABI version on *nix compatible systems is weird. Usually the soversion is a monotonically non-decreasing integer, that's why it seems weird to me. Maybe an additional variable is in order right next to `VERSION_IS_RELEASE` like `set(SFML_ABI_SOVERSION ...)`?

> I suggest to start with low-hanging fruits (e.g. proper use of CMake-options with ON/OFF or similar) and ensure that we don't get huge diffs

I'm conflicted on this. Generally, I agree that huge PRs can surely be broken down, but with the amount of research and considering I have been mostly removing code so far, I'm not sure it's worth doing this in a piecemeal fashion. Most of the additions would involve better documentation surrounding the build process anyway. I'll see what the end result will look like, I also have a couple notes that I would like to get answers to once I have a draft PR up.

**Comment by ChrisThrasher** [comment]
>  I have already chucked quite a bit out the window, since a lot of it is simply redundant with higher CMake versions, or simply there was no need for it to begin with.

Can these removals be PRs on their own? This is exactly what everybody is most eager to see PRs for.

**Comment by friendlyanon** [comment]
Okay, so I have been chipping away at this things and - owning no Apple products - I am having trouble researching how frameworks and bundles fit into the big picture.

CMake offers a way to specify the `DESTINATION` of `BUNDLE`s and `FRAMEWORK`s, but without defaults. I am curious what destination would be idiomatic for the platform and what destination other things should get in relation. It's extra difficult figuring this out without having an Apple machine and I can't even run macos in a VM, because it makes my laptop keel over with the CPU usage.

CMake's test suite has a single test that tests installing frameworks and the destination there is just [hardcoded to `lib`](https://gitlab.kitware.com/cmake/cmake/-/blob/master/Tests/RunCMake/XcodeProject/XcodeIOSInstallCombined.cmake#L36).

I tried looking at git blame for the relevant lines of code and I found no information that would help here either. PR that added framework support: #12

**Comment by friendlyanon** [comment]
I have a WIP branch up for the curious ones: https://github.com/friendlyanon/SFML/tree/feature/cmake-rework-wip  
`48 files changed, 956 insertions(+), 2856 deletions(-)` if you don't count the deletion of vendored dependencies and there will be more deletions after it's done.

The library builds, as static or shared, and it has a proper install interface that propagates transitive requirements.  
I ripped out the pkg-config code, as it's not easily supported for non-trivial cases in CMake.

Provided you have system dependencies like the Vulkan SDK, X11 or `libdrm-dev` and `libgbm-dev` on Linux, you can build the WIP branch with vcpkg as it is. If you want to use all system dependencies, then the only thing to look out for is that only OpanAL Soft 1.21+ works, because that's the version that started providing support for CMake clients, although the CMake package is broken due to a missing version script, so you can't require that version in the CMake code. I left a comment above the `find_package` call about this.

I have tests, docs and examples left to rework, and of course the framework question from above.

**Comment by friendlyanon** [comment]
I have updated the above WIP branch and it's nearly finished now, minus the unanswered bits. Pretty much all that's left is to slice the mega commit apart, add presets, update CI workflows and docs.

**Comment by ChrisThrasher** [comment]
Are you interested in submitting smaller PRs for any of the work you've done in your cmake-rework-wip branch? You've spoken a few times about places where we can remove code in favor of using built-in CMake feature and I'm still curious what you were specifically referencing.

**Comment by ChrisThrasher** [comment]
If you would like to continue talking about this, please do so in #1991.
